core/src/main/java/org/elasticsearch/rest/action/support/RestToXContentListener.java

```diff
@@ -45,6 +45,10 @@ public final RestResponse buildResponse(Response response, XContentBuilder build
         builder.startObject();
         response.toXContent(builder, channel.request());
         builder.endObject();
-        return new BytesRestResponse(RestStatus.OK, builder);
+        return new BytesRestResponse(getStatus(response), builder);
```

Comment:
This seemed like a simple place to implement request status. I'm frankly surprised no one did this before. Anyway, this is exactly how it works in master.